Dependencies/xr/Source/ARKit/XR.mm

```diff
@@ -0,0 +1,462 @@
+                } RasterizerData;
+
+                vertex RasterizerData
+                vertexShader(uint vertexID [[vertex_id]],
+                             constant XRVertex *vertices [[buffer(0)]])
+                {
+                    RasterizerData out;
+                    out.position = vector_float4(vertices[vertexID].position.xy, 0.0, 1.0);
+                    out.uv = vertices[vertexID].uv;
+                    return out;
+                }
+
+                fragment float4 fragmentShader(RasterizerData in [[stage_in]],
+                    texture2d<half> babylonTexture [[ texture(0) ]],
+                    texture2d<half> cameraTexture [[ texture(1) ]])
+                {
+                    constexpr sampler linearSampler(mag_filter::linear, min_filter::linear);
+
+                    const half4 babylonSample = babylonTexture.sample(linearSampler, in.uv);
+                    const half4 cameraSample = cameraTexture.sample(linearSampler, in.uv);
+                    const half4 mixed = mix(cameraSample, babylonSample, babylonSample.a);
+                    return vector_float4(mixed);
+                }
+            )";
+            
+            id<MTLLibrary> lib = CompileShader(source);
+            id<MTLFunction> vertexFunction = [lib newFunctionWithName:@"vertexShader"];
+            id<MTLFunction> fragmentFunction = [lib newFunctionWithName:@"fragmentShader"];
+
+            // Configure a pipeline descriptor that is used to create a pipeline state.
+            MTLRenderPipelineDescriptor *pipelineStateDescriptor = [[MTLRenderPipelineDescriptor alloc] init];
+            pipelineStateDescriptor.label = @"XR Pipeline";
+            pipelineStateDescriptor.vertexFunction = vertexFunction;
+            pipelineStateDescriptor.fragmentFunction = fragmentFunction;
+            pipelineStateDescriptor.colorAttachments[0].pixelFormat = MetalLayer.pixelFormat;
+
+            pipelineState = [MetalDevice newRenderPipelineStateWithDescriptor:pipelineStateDescriptor
+                                                                     error:&error];
+            if (!pipelineState)
+            {
+                NSLog(@"Failed to create pipeline state: %@", error);
+            }
+            commandQueue = [MetalDevice newCommandQueue];
+            
+            auto scale = UIScreen.mainScreen.scale;
+            viewportSize.x = MetalLayer.visibleRect.size.width * scale;
+            viewportSize.y = MetalLayer.visibleRect.size.height * scale;
+        }
+
+        ~Impl()
+        {
+            [sessionDelegate release];
+            [configuration release];
+            [session release];
+        }
+
+        id<MTLLibrary> CompileShader(const char* source)
+        {
+            NSError* error;
+            id<MTLLibrary> lib = [MetalDevice newLibraryWithSource:@(source) options:nil error:&error];
+            if(NULL != error)
+            {
+				NSLog(@"Shader compilation failed: %s"
```

Comment:
Should we throw an exception? This is what we do on the ARCore side, seems like it would be good to have both behave the same.

Reply:
Done.